test/unit/test_keyring_multi.py

```diff
@@ -0,0 +1,242 @@
+    )
+
+
+def test_identity_keyring_as_generator_and_no_data_encryption_key():
+    test_multi_keyring = MultiKeyring(generator=IdentityKeyring())
+    with pytest.raises(GenerateKeyError) as exc_info:
+        test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_without_data_key())
+    assert exc_info.match("Unable to generate data encryption key.")
+
+
+def test_number_of_encrypted_data_keys_without_generator_with_children():
+    test_multi_keyring = get_multi_keyring_with_no_generator()
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    assert len(test.encrypted_data_keys) == len(test_multi_keyring.children)
+
+
+def test_number_of_encrypted_data_keys_without_children_with_generator():
+    test_multi_keyring = get_multi_keyring_with_no_children()
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    assert len(test.encrypted_data_keys) == 1
+
+
+def test_number_of_encrypted_data_keys_with_generator_and_children():
+    test_multi_keyring = get_multi_keyring_with_generator_and_children()
+    number_of_children = len(test_multi_keyring.children)
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    assert len(test.encrypted_data_keys) == number_of_children + 1
+
+
+def test_on_encrypt_when_data_encryption_key_given():
+    test_multi_keyring = MultiKeyring(generator=mock_generator, children=[mock_child_1, mock_child_2])
+    test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_key())
+    for keyring in test_multi_keyring._decryption_keyrings:
+        keyring.on_encrypt.assert_called_once()
+
+
+def test_on_encrypt_edk_length_when_keyring_generates_but_does_not_encrypt():
+    test_multi_keyring = MultiKeyring(generator=OnlyGenerateKeyring())
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_without_data_key())
+    assert test.data_encryption_key is not None
+    assert len(test.encrypted_data_keys) == len(get_encryption_materials_without_data_key().encrypted_data_keys)
+
+    test = test_multi_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_encrypted_data_key())
+    assert len(test.encrypted_data_keys) == len(get_encryption_materials_with_encrypted_data_key().encrypted_data_keys)
+
+
+def test_on_decrypt_when_data_encryption_key_given():
+    test_multi_keyring = MultiKeyring(generator=mock_generator, children=[mock_child_1, mock_child_2])
+    test_multi_keyring.on_decrypt(decryption_materials=get_decryption_materials_with_data_key(), encrypted_data_keys=[])
+    for keyring in test_multi_keyring._decryption_keyrings:
+        keyring.on_decrypt.assert_not_called()
+
+
+def test_on_decrypt_every_keyring_called_when_data_encryption_key_not_added():
+    mock_generator.on_decrypt.return_value = get_decryption_materials_without_data_key()
```

Comment:
Let's be a bit less brittle and decouple these from what we assume the input will be:
```suggestion
    mock_generator.on_decrypt.side_effect = lambda x: x
```

This changes the call behavior of `mock_generator.on_decrypt` to be the identity function, but still retains the mock instrumenting.